File: src/16_3Sum_Closest.rs
// 16_3Sum_Closest
// https://leetcode.cn/problems/3sum-closest/

impl Solution {
    pub fn three_sum_closest(nums: Vec<i32>, target: i32) -> i32 {
        let mut nums = nums;
        let len = nums.len();
        if len == 3 {
            return nums.iter().sum::<i32>();
        }

        nums.sort_unstable();
        
        let m = nums[0..3].iter().sum();
        if m >= target {
            return m;
        }
        let x = nums[nums.len() - 3..nums.len()].iter().sum();
        if x <= target {
            return x;
        }

        let mut diff = i32::MAX;
        for i in 0..len-2 {
            let t = target - nums[i];
            let mut left = i+1;
            let mut right = len-1;
            while left < right {
                if nums[left] + nums[right] < t {
                    if (nums[left] + nums[right] - t).abs() < diff.abs() {
                        diff = nums[left] + nums[right] - t;
                    }
                    left += 1;
                } else if nums[left] + nums[right] > t {
                    if (nums[left] + nums[right] - t).abs() < diff.abs() {
                        diff = nums[left] + nums[right] - t;
                    }
                    right -= 1
                } else {
                    return target;
                }
            }
        }
        return target + diff

    }
}


impl Solution {
    pub fn three_sum_closest(mut nums: Vec<i32>, target: i32) -> i32 {
        nums.sort_unstable();

        let m = nums[0..3].iter().sum();
        if m >= target {
            return m;
        }
        let x = nums[nums.len() - 3..nums.len()].iter().sum();
        if x <= target {
            return x;
        }

        let mut c = if target - m > x - target { m } else { x };
        for i in 0..nums.len() - 1 {
            let (mut l, mut r) = (i + 1, nums.len() - 1);
            while l < r {
                let s = nums[i] + nums[l] + nums[r];
                if s < target {
                    if target - s < (target - c).abs() {
                        c = s;
                    }
                    l += 1;
                } else if s > target {
                    if s - target < (c - target).abs() {
                        c = s;
                    }
                    r -= 1;
                } else {
                    return target;
                }
            }
        }

        c
    }
}

File: src/16_3Sum_Closest.py
# 16_3Sum_Closest
# https://leetcode.cn/problems/3sum-closest/

# 给你一个长度为 n 的整数数组 nums 和 一个目标值 target。请你从 nums 中选出三个整数，使它们的和与 target 最接近。
# 返回这三个数的和。
# 假定每组输入只存在恰好一个解。

# 示例 1：
# 输入：nums = [-1,2,1,-4], target = 1
# 输出：2
# 解释：与 target 最接近的和是 2 (-1 + 2 + 1 = 2) 。

# 示例 2：
# 输入：nums = [0,0,0], target = 1
# 输出：0


from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        n = len(nums)
        best = 10**7
        
        # 根据差值的绝对值来更新答案
        def update(cur):
            nonlocal best
            if abs(cur - target) < abs(best - target):
                best = cur
        
        # 枚举 a
        for i in range(n):
            # 保证和上一次枚举的元素不相等
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            # 使用双指针枚举 b 和 c
            j, k = i + 1, n - 1
            while j < k:
                s = nums[i] + nums[j] + nums[k]
                # 如果和为 target 直接返回答案
                if s == target:
                    return target
                update(s)
                if s > target:
                    # 如果和大于 target，移动 c 对应的指针
                    k0 = k - 1
                    # 移动到下一个不相等的元素
                    while j < k0 and nums[k0] == nums[k]:
                        k0 -= 1
                    k = k0
                else:
                    # 如果和小于 target，移动 b 对应的指针
                    j0 = j + 1
                    # 移动到下一个不相等的元素
                    while j0 < k and nums[j0] == nums[j]:
                        j0 += 1
                    j = j0

        return best  

# 有两个优化可以让代码击败接近 100%！
# 设 s = nums[i] + nums[i+1] + nums[i+2]。
# 如果 s > target，由于数组已经排序，后面无论怎么选，选出的三个数的和不会比 s 还小，所以不会找到比 s 更优的答案了。
# 所以只要 s > target，就可以直接 break 外层循环了。
# 在 break 前判断 s 是否离 target 更近，如果更近，那么更新答案为 s。

# 设 s = nums[i] + nums[n-2] + nums[n-1]。如果 s < target，
# 由于数组已经排序，nums[i] 加上后面任意两个数都不超过 s，所以下面的双指针就不需要跑了，
# 无法找到比 s 更优的答案。但是后面还有更大的 nums[i]，可能找到一个离 target 更近的三数之和，
# 所以还需要继续枚举，continue 外层循环。
# 在 continue 前判断 s 是否离 target 更近，如果更近，那么更新答案为 s。

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        n = len(nums)
        min_diff = float("inf")
        for i in range(n - 2):
            x = nums[i]
            if i and x == nums[i - 1]:
                continue  # 优化三

            # 优化一
            s = x + nums[i + 1] + nums[i + 2]
            if s > target:  # 后面无论怎么选，选出的三个数的和不会比 s 还小
                if s - target < min_diff:
                    ans = s
                break

            # 优化二
            s = x + nums[-2] + nums[-1]
            if s < target:  # x 加上后面任意两个数都不超过 s，所以下面的双指针就不需要跑了
                if target - s < min_diff:
                    min_diff = target - s
                    ans = s
                continue

            # 双指针
            j, k = i + 1, n - 1
            while j < k:
                s = x + nums[j] + nums[k]
                if s == target:
                    return s
                if s > target:
                    if s - target < min_diff:  # s 与 target 更近
                        min_diff = s - target
                        ans = s
                    k -= 1
                else:  # s < target
                    if target - s < min_diff:  # s 与 target 更近
                        min_diff = target - s
                        ans = s
                    j += 1
        return ans



# GPT-4
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        closest_sum = float('inf')

        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left, right = i + 1, len(nums) - 1
            while left < right:
                current_sum = nums[i] + nums[left] + nums[right]
                
                if current_sum == target:
                    return current_sum

                if abs(current_sum - target) < abs(closest_sum - target):
                    closest_sum = current_sum

                if current_sum < target:
                    left += 1
                elif current_sum > target:
                    right -= 1

        return closest_sum